### PR TITLE
[gh-24339] Move from streaming stats to polling for docker

### PR DESCRIPTION
**Pull request by Juanadelacuesta**
### Description
It addresses [24339](https://github.com/hashicorp/nomad/issues/24339). 

Currently to get the stats from a container we are using the  call to `ContainerStats` which returns a stream of data that can be reused, but it only works if the sampling interval for the metrics is `1s` exactly. With other values, decoding errors start to show up and the metric values becomes inconsistent:

```
2024-11-21T13:14:17.645+0100 [DEBUG] client.driver_mgr.docker: error decoding stats data from container: container_id=2317edae0f7a0e1fb2e259a2e87f2575cf6114a2c2c47c3eaf7f992de0b7a569 driver=docker error="json: cannot unmarshal string into Go value of type container.Stats"
2024-11-21T13:14:17.651+0100 [DEBUG] client.driver_mgr.docker: error decoding stats data from container: container_id=f062a6097af27095c1e8eec76005d33c670edaadb2e5b74540f83704f0260a60 driver=docker error="invalid character 'a' looking for beginning of value"
```
It might have gone unnoticed because the error messages were displayed only for debug, which rarely happens on production environments.

To avoid the sync problem, this PR introduces a polling call to the stats endpoint, which is not as efficient, but is more versatile and allows for different sampling intervals.

### Testing & Reproduction steps
To see the bug, take the latest release and start a nomad agent, it can be in -dev mode, with telemetry enabled and a `collection_interval` different from `1s`.

```
telemetry {
  collection_interval        = "3s"
  prometheus_metrics         = true
  publish_allocation_metrics = true
}
```
Run a job with multiple allocations and after a while the errors will show and the metrics for the allocations will be unstable. 

### Contributor Checklist
- [ ] **Changelog Entry** If this PR changes user-facing behavior, please generate and add a
  changelog entry using the `make cl` command.
- [ ] **Testing** Please add tests to cover any new functionality or to demonstrate bug fixes and
  ensure regressions will be caught.
- [ ] **Documentation** If the change impacts user-facing functionality such as the CLI, API, UI,
  and job configuration, please update the  Nomad website documentation to reflect this. Refer to
  the [website README](../website/README.md) for docs guidelines. Please also consider whether the
  change requires notes within the [upgrade guide](../website/content/docs/upgrade/upgrade-specific.mdx).

### Reviewer Checklist
- [x] **Backport Labels** Please add the correct backport labels as described by the internal
  backporting document.
- [x] **Commit Type** Ensure the correct merge method is selected which should be "squash and merge"
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
- [ ] **Enterprise PRs** If this is an enterprise only PR, please add any required changelog entry
  within the public repository.

### CODE REVIEW

File: drivers/docker/stats.go [hunk 1]
```diff
@@ -16,6 +16,8 @@ import (
 	cstructs "github.com/hashicorp/nomad/client/structs"
 	"github.com/hashicorp/nomad/drivers/docker/util"
 	"github.com/hashicorp/nomad/helper"
+
+	//"github.com/hashicorp/nomad/helper"
```
Comment (tgross) [inline drivers/docker/stats.go:20]: Let's clean this up.

**Review by tgross — COMMENTED** [review]
LGTM in terms of implementation, but can we describe how this has been tested in this PR? The original fix to use streaming during the 1.9.0-beta was to fix a leaked goroutine, so I really want to make sure we're not regressing on problems like that.

**Comment by Juanadelacuesta** [comment]
> LGTM in terms of implementation, but can we describe how this has been tested in this PR? The original fix to use streaming during the 1.9.0-beta was to fix a leaked goroutine, so I really want to make sure we're not regressing on problems like that.

Do you mean tests in terms of no leaks? Maybe there was a previous PR touching this same code, but the latest was about using a different call to get a more complete stats report from docker.

**Comment by tgross** [comment]
> Do you mean tests in terms of no leaks? Maybe there was a previous PR touching this same code, but the latest was about using a different call to get a more complete stats report from docker.

Right, I'm just asking that you verified this doesn't leak any goroutines from the called functions. You can do it manually, I just want to make sure we've done so, so that we don't have a regression.

**Comment by Juanadelacuesta** [comment]
I did not, but I will non

**Comment by Juanadelacuesta** [comment]
Tests to detect leaking go routines were manually run using [`goroutine-inspect`](https://github.com/tgross/goroutine-inspect) by checking for go routines before, during and after running a docker workload with multiple allocations and verifying the go back to the initial number.

**Review by tgross — APPROVED** [review]
LGTM!

**Comment by Himura2la** [comment]
Does it really address #24339 ?

**Pull request merged**